Boss Logger Plugin

TheStonedTurtle opened the pull request:
Records loot received from boss monsters in a `.log` file inside the .runelite directory.

* Tracks loot received from the boss whether you looted it or not.
* Has a panel showing all loot and the price of each item, see picture(s) below.
* Tracks Unique Items from each boss within panel, see picture(s) below.
* In-game and system notification alerts with toggle options
* Records the following bosses:
	* Armadyl 
	* Abyssal Sire
	* Bandos - TESTED
	* Barrows Chests - TESTED
	* Callisto
	* Cerberus
	* Chaos Elemental - TESTED
	* Chaos Fanatic
	* Corporeal Beast
	* Crazy Archaeologist - TESTED
	* Dag Rex
	* Dag Prime
	* Dag Supreme
	* Giant Mole - TESTED
	* Grotesque Guardians - TESTED
	* Kalphite Queen - TESTED
	* King black dragon - TESTED
	* Kraken
	* Saradomin - TESTED
	* Scorpia - TESTED
	* Skotizo - TESTED
	* Raids
	* Thermonuclear Smoke Devil
	* Vet'ion
	* Venenatis
	* Vorkath - TESTED
	* Zammy - TESTED
	* Zulrah - TESTED

I still need to test some bosses and the tracking system for exchanging sires at the pool. The list above has the currently tested bosses, if anyone could help with testing that would be greatly appreciated.

The goal is to eventually use the functionality from @WooxSolo's droplogger for finding the correct loot tile but I have a feeling that it may be implemented as part of the API. Because of this some of the loot finding code has been pulled from Woox's plugin.

Panel Gif:
![2018-05-24_03-56-20](https://user-images.githubusercontent.com/29030969/40814228-ddd6a25c-64f3-11e8-95b8-258e9916d43e.gif)

Panel with some uniques received:
![javaw_2018-05-31_16-48-45](https://user-images.githubusercontent.com/29030969/40814213-c2a51edc-64f3-11e8-9383-0fa8b7214416.png)

Example In-game message:
![java_2018-05-28_13-23-01](https://user-images.githubusercontent.com/29030969/40814260-0b449d02-64f4-11e8-8ffe-105501e9c60d.png)


NOTES: 
* This data is stored in a folder with the same name as your ingame username. When you change your username you will need to manually update this folder or a new folder will be created.
* Some of the bosses were tested prior to the removal of `onActorDeath` but i believe my work around will not affect the loot tile tracking system.
* This is my firt time programming in Java so there is probably some code simplifications/optimizations that can be made.

Legendarybotter commented:
Love this plugin <3

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/BossLoggerPanel.java:
```diff
@@ -0,0 +1,267 @@
+		scroller.setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_NEVER);
+		scroller.getVerticalScrollBar().setUnitIncrement(16);
+
+
```
Remove empty line

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/BossLoggerPanel.java:
```diff
@@ -0,0 +1,267 @@
+				}
+			}
+		}
+		return insertTabAt;
```
Add empty line above

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/BossLoggerPanel.java:
```diff
@@ -0,0 +1,267 @@
+				tabPositions.put(entry.getKey(), entry.getValue() - 1);
+			}
+		}
+		tabPositions.remove(tabName);
```
Add line above

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/DropEntry.java:
```diff
@@ -0,0 +1,42 @@
+	private final Integer item_id;
+
+	@Getter
+	private final Integer item_amount;
```
Why are the variable types Integer instead of int? Also, have variable declarations on the top of the class, above the constructor.

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/DropEntry.java:
```diff
@@ -0,0 +1,42 @@
+
+class DropEntry
+{
+	DropEntry(Integer item_id, Integer item_amount)
```
Use camelCase for variable names:
itemId
itemAmount

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/LootEntry.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.ArrayList;
+import lombok.Getter;
+
+class LootEntry
```
Consider the same review as the DropEntry class, variable types, naming, and placement.

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/LootPanel.java:
```diff
@@ -0,0 +1,207 @@
+		c.gridx = 0;
+		c.gridy = 0;
+
+
```
Remove empty line

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/LootRecord.java:
```diff
@@ -0,0 +1,83 @@
+	@Setter(AccessLevel.PACKAGE)
+	private ItemComposition item;
+
+	LootRecord(String item_name, Integer item_id, Integer amount, Integer value, AsyncBufferedImage icon, ItemComposition item)
```
Consider the same review as the DropEntry class, variable types, naming and placement.

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/LootRecord.java:
```diff
@@ -0,0 +1,83 @@
+		this.value = value;
+
+		long val = this.value;
+		long am = this.amount;
```
Why don't you just use the original variables instead of creating new ones?

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/LootRecord.java:
```diff
@@ -0,0 +1,83 @@
+	{
+		l.amount = l.amount + amount;
+		long val = this.value;
+		long am = this.amount;
```
Same here, why create new variables?

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/LootRecord.java:
```diff
@@ -0,0 +1,83 @@
+
+	void updateIconAmount(LootRecord l, ItemManager itemManager)
+	{
+		Boolean stackable = l.item.isStackable() || l.amount > 1;
```
Why are you using the wrapper classes? use the primitive boolean type.

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/LootRecordPanel.java:
```diff
@@ -0,0 +1,162 @@
+		c.gridx = 0;
+		c.gridy = 0;
+		c.ipady = 20;
+
```
Remove empty line

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/LootRecordPanel.java:
```diff
@@ -0,0 +1,162 @@
+	}
+
+	// Add K and M Suffix to items
+	private String numberToString(long v)
```
Use the StackFormatter class

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/LootRecordPanel.java:
```diff
@@ -0,0 +1,162 @@
+	}
+
+	// Color label to match RuneScape coloring
+	private void ColorLabel(JLabel label, long val)
```
this is the wrong naming for methods, use camelCase.

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/LootRecordPanel.java:
```diff
@@ -0,0 +1,162 @@
+		{
+			labelColor = Color.white;
+		}
+		label.setForeground(labelColor);
```
Also this methodcould be reduced by a bunch of lines by using a ternary operator.

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/UniqueItemPanel.java:
```diff
@@ -0,0 +1,138 @@
+	private ArrayList<UniqueItem> items;
+	private Map<String, LootRecord> loots;
+
+	private final Float alpha_missing = 0.3f;
```
Use float, the primitive type.

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/UniqueItemPanel.java:
```diff
@@ -0,0 +1,138 @@
+	private final Float alpha_missing = 0.3f;
+	private final Float alpha_has = 1.0f;
+
+	UniqueItemPanel(ArrayList<UniqueItem> _items, Map<String, LootRecord> _loots, ItemManager _itemManager)
```
Use cameCase.

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/UniqueItemPanel.java:
```diff
@@ -0,0 +1,138 @@
+			};
+			image.onChanged(() -> SwingUtilities.invokeLater(task));
+		}
+
```
Remove empty line

psikoi commented:
Would it be possible for me to manually add items? I have a similar spreadsheet with drops that id like to transfer over once this is done

TheStonedTurtle commented:
@psikoi Yeah you can manually add them in, I did that for about 225 of my barrows entries. 

In `.runelite/loots/` there will be a folder with your username and inside of there it will have all of the log files. There is 1 file for each boss and the file gets created after the first kill with the plugin enabled.

Example Format:
`{"killCount":1,"drops":[{"itemAmount":1536,"itemId":995},{"itemAmount":36,"itemId":4740}]}`

Each entry should be on a separate line inside the file, any blank lines will be ignored. The whole reading/writing system may be updated to be more efficient/output a different format after review.

Here is my `barrows.log` file via pastebin:
https://pastebin.com/LR1UR9qi

psikoi commented:
I wonder if this plugin can be refactored/rebranded as a more generic collection plugin. Id likr to have clue items, untradables ive gotten (void, f cape, etc). I have a photoshop file wih a bunch of item icons that shows me my unlocked items, i think it could be good

TheStonedTurtle commented:
The clue items would be super easy to add, like most other items that are received from drops. Void would require some figuring out since you exchange points for it.

I have no problem with this be refactored/rebranded i just thought tracking all the bosses was a good first step for an Iron Man Tracker.

psikoi commented:
Maybe void was a bad example, but some things that would fit nicely besides clues is like pharao's sceptre, etc.. I think movinf away from just bosses would be good

Jalopyy commented:
Should be nice to see this extended to a slayer logger plugin. Sometimes I do slayer at places like money monkeys or wonder how much profit I got from doing gargs.

Jeremy-Excersion commented:
I like the plugin, although a dropdown menu might be easier to select a specific boss, rather than the boss icon and left/right arrows.

TheStonedTurtle commented:
## Changes
* Updated Panel UI for better Tab Selection (See Media Below)
* Added Raid 2 Tab, currently not recording loot due to not having chest inventory/widget ID.
* Automatically selects boss tab on kill if still on selection screen (See Media Below)
* Added ability to clear stored data with confirmation screen


## Media
<details>
<p>

### New UI
![2018-06-16_04-56-30](https://user-images.githubusercontent.com/29030969/41539147-e14f2fb8-72c1-11e8-81ab-e8c5617211a1.gif)
(The empty border on the right side has been fixed when no scroll bar, didn't want to remake a gif)

## Auto select boss panel functionality
![2018-06-18_06-24-40](https://user-images.githubusercontent.com/29030969/41539117-d04a404a-72c1-11e8-9260-d10716030c4e.gif)

</p>
</details>

NathenSample commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/BossLoggerPlugin.java:
```diff
@@ -0,0 +1,1365 @@
+		init();
+
+		// Ensure Loot Directory has been created
+		LOOTS_DIR.mkdir();
```
Worth verifying the return of this in order to handle a scenario where creating the directory failed

NathenSample commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/BossLoggerPanel.java:
```diff
@@ -0,0 +1,437 @@
+			return;
+		}
+
+		int delete = JOptionPane.showConfirmDialog(this.getRootPane(), "<html>Are you sure you want to clear all data for this tab?<br/>There is no way to undo this action.</html>", "Warning", JOptionPane.YES_NO_OPTION);
```
May be worth considering moving this, and any other HTML into either class level finals, or a HTML const file

NathenSample commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/BossLoggerPanel.java:
```diff
@@ -0,0 +1,437 @@
+			}
+			catch (IOException e)
+			{
+				log.warn("Error getting resource icon: " + iconName, e);
```
This is only a warn, and the catch performs no other actions, is a full stack trace necessary or could it be limited to `e.getMessage();`?

NathenSample commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/BossLoggerPlugin.java:
```diff
@@ -0,0 +1,1365 @@
+		String chatMessage = event.getMessage();
+
+		// Barrows KC
+		if (chatMessage.startsWith("Your Barrows chest count is"))
```
Rather than doing a string comparison here, you could instead use a regular expression as you do below that will capture the kill value - If you get a returned value then you know the chat message started in the correct form.

Minor, but means you just apply a regex, rather than string comparison and a regex.

NathenSample commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/BossLoggerPlugin.java:
```diff
@@ -0,0 +1,1365 @@
+			// Is Boss being recorded?
+			mapRecording.put(bossName, isBeingRecorded(tab.getName()));
+			// Loot Entries by Tab Name
+			ArrayList<LootEntry> array = new ArrayList<LootEntry>();
```
Nitpicking but declare the interface, provide an implementation

`List<LootEntry> array = new ArrayList<LootEntry>();`

NathenSample commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/BossLoggerPlugin.java:
```diff
@@ -0,0 +1,1365 @@
+			}
+			catch (IOException e)
+			{
+				log.warn("Error getting panel icon:", e);
```
Again a warn with no other action, I'll stop commenting on these, but it's always worth considering whether the log at any given point needs the stack trace, or the message will suffice. 

It's not wrong, it's just a worthwhile consideration.

NathenSample commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/BossLoggerPlugin.java:
```diff
@@ -0,0 +1,1365 @@
+		String fileName = filenameMap.get(bossName.toUpperCase());
+		// Open File and append data
+		File lootFile = new File(playerFolder, fileName);
+		try
```
This try should probably have a finally to close the file handle, i'd recommend looking into [try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html)

NathenSample commented on runelite-client/src/main/java/net/runelite/client/plugins/bosslogger/BossLoggerPlugin.java:
```diff
@@ -0,0 +1,1365 @@
+		String fileName = filenameMap.get(tab.getBossName().toUpperCase());
+		File lootFile = new File(playerFolder, fileName);
+
+		try
```
Same as the other try, if you're using resources make sure to close them in a finally, or try with
